### PR TITLE
Adding basic unit tests for wazp.utils.

### PR DESCRIPTION
**Pull request by samcunliffe**
More noise from me.

Here is a (very) basic unit test for `wazp.utils.df_from_metadata_yaml_files` (relates to #27).

Now [tox py38 (pytest@3.8) fails](https://github.com/SainsburyWellcomeCentre/WAZP/actions/runs/4182565018/jobs/7245871921#step:2:500) because of the `TypeError` due to [this deprecation](https://docs.python.org/3/library/typing.html?highlight=subscripting#typing.List), or rather: due to the extension of `builtin` which happens 3.8 → 3.9. (We're using the 3.9+ standard 👍.) 

~~I'm not entirely sure why we only notice this now.~~ ... oh: because we've probably never tried running WAZP in python3.8.

### CODE REVIEW

File: tests/test_unit/test_utils.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+
+def test_df_from_metadata_yaml_sample_project_metadata() -> None:
+    """Normal operation: test we can read the sample project metadata."""
+    fields = get_sample_project_metadata_fields()
+    testme = df_from_metadata_yaml_files("sample_project/videos", fields)
+
+    expected = set(fields)
+    actual = set(testme.columns)
+    diff = expected.symmetric_difference(actual)
```
Comment (sfmig) [inline tests/test_unit/test_utils.py:24]: should we compute diff only if the don't match?

File: tests/test_unit/test_utils.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    return metadata_fields
+
+
+def test_df_from_metadata_yaml_sample_project_metadata() -> None:
+    """Normal operation: test we can read the sample project metadata."""
+    fields = get_sample_project_metadata_fields()
+    testme = df_from_metadata_yaml_files("sample_project/videos", fields)
+
+    expected = set(fields)
+    actual = set(testme.columns)
```
Comment (sfmig) [inline tests/test_unit/test_utils.py:23]: I'd suggest more descriptive names for the variables, but this might be personal:
fields --> fields_dict, or metadata_yaml?
testme ---> df_output
expected --> fields_from_yaml
actual ---> df_columns

File: tests/test_unit/test_utils.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+
+def get_sample_project_metadata_fields() -> dict:
+    """Get the metadata dictionary from the sample project for testing."""
+    with open("sample_project/metadata_fields.yaml") as fi:
+        metadata_fields = yaml.safe_load(fi)
+    return metadata_fields
+
+
+def test_df_from_metadata_yaml_sample_project_metadata() -> None:
```
Comment (sfmig) [inline tests/test_unit/test_utils.py:17]: not sure if this is against convention but maybe this would be more descriptive:
'test_columns_names_and_nrows_in_df_from_metadata'

File: tests/test_unit/test_utils.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    """Normal operation: test we can read the sample project metadata."""
+    fields = get_sample_project_metadata_fields()
+    testme = df_from_metadata_yaml_files("sample_project/videos", fields)
+
+    expected = set(fields)
+    actual = set(testme.columns)
+    diff = expected.symmetric_difference(actual)
+    assert expected == actual, f"Metadata fields -> df problem with: {diff}"
```
Comment (sfmig) [inline tests/test_unit/test_utils.py:25]: I'd rephrase this a bit:
f"Metadata fields and df columns differ in the following fields: {diff}"

Or maybe more explicitly:
f"The following fields appear only in either the metadata fields or the df columns: {diff}"

File: tests/test_unit/test_utils.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    assert nrows == nfiles, "Number of rows in df != number of yaml files."
+
+
+def test_df_from_metadata_yaml_no_metadata() -> None:
+    """
+    Test with no metadata files (expect just to create an empty dataframe with
+    metadata_fields column headers).
+    """
+    fields = get_sample_project_metadata_fields()
+    with tempfile.TemporaryDirectory() as empty_existing_directory:
```
Comment (sfmig) [inline tests/test_unit/test_utils.py:38]: nice! ✨ 
I wasn't aware of this way of making temp dirs

**Comment by sfmig** [comment]
A more general question: should we add a check in the definition of the `df_from_metadata_yaml_files` function, that throws an error if the metadata fields from the yaml file and the column names don't match? Since I guess here we are only testing for the sample project.... 🤔

**Review by sfmig — APPROVED** [review]

**Comment by niksirbi** [comment]
> A more general question: should we add a check in the definition of the `df_from_metadata_yaml_files` function, that throws an error if the metadata fields from the yaml file and the column names don't match? Since I guess here we are only testing for the sample project....

At some point we should definitely add this. On the same topic, now that I started adding the ROI coordinates to the metadata file, I stumbled into problems. I added the relevant description in `metadata_fields.yaml` but that was not enough. Due to the way the metadata table is currently built, it cannot handle the fact that some metadata files contain a field (e.g. "ROI_coords"), while others do not. I currently went around this issue for the sake of the ROI PR (by basically building the table using only shared columns), but we should definitely make the process robust to such things.

**Comment by sfmig** [comment]
Re: tox error
Shall we remove py3.8 then?

File: tests/test_unit/test_utils.py (hunk 5 above)
Comment (niksirbi) [reply tests/test_unit/test_utils.py:38]: I independently stumbled upon this yesterday, very neat!

**Comment by niksirbi** [comment]
> Now [tox py38 (pytest@3.8) fails](https://github.com/SainsburyWellcomeCentre/WAZP/actions/runs/4182565018/jobs/7245871921#step:2:500) because of the `TypeError` due to [this deprecation](https://docs.python.org/3/library/typing.html?highlight=subscripting#typing.List), or rather: due to the extension of `builtin` which happens 3.8 → 3.9. (We're using the 3.9+ standard .)

Could this be solved by using say `typing.List` instead of `list`?

**Comment by sfmig** [comment]
`typing.List` is [deprecated](https://docs.python.org/3/library/typing.html?highlight=subscripting#typing.List) 
> Could this be solved by using say `typing.List` instead of `list` in the type hint?

File: tests/test_unit/test_utils.py (hunk 5 above)
Comment (samcunliffe) [reply tests/test_unit/test_utils.py:38]: 😎

**Comment by niksirbi** [comment]
> typing.List is [deprecated](https://docs.python.org/3/library/typing.html?highlight=subscripting#typing.List) (at least from 3.9)!

That's unfortunate. So `typing.List` (and `Dict` etc.) does not work for >= 3.9 and `list[str]` does not work for < 3.9.

I am not keen to drop 3.8 support easily (if it can be avoided in some way). If people would want to use this package in the same environment as say `deeplabcut`, that could cause conflicts.

File: tests/test_unit/test_utils.py (hunk 1 above)
Comment (samcunliffe) [reply tests/test_unit/test_utils.py:24]: For efficiency? I was wondering about that too but I couldn't come up with a snappy way to do it. [According to the python wiki `symmetric_difference` is only O(n)](https://wiki.python.org/moin/TimeComplexity) and we're not going crazy with metadata, right? So n ~ 30.

**Comment by niksirbi** [comment]
> I don't think building the table using only shared columns is a good solution, and I don't think that the metadata files per video having different fields should be allowed. It makes sense that if each video metadata file represents a row, then all rows have the same fields (i.e., column names). So I think it is better if all video metadata files have an ROI field, even if this is empty.

I am fine with this solution: requiring all metadata files to have the specified fields, but allowing the values to be empty/nan.
If the `roi_coords` field ends up being ugly in the Dash DataTable, we could hide it by default.

**Comment by samcunliffe** [comment]
>> typing.List is [deprecated](https://docs.python.org/3/library/typing.html?highlight=subscripting#typing.List) (at least from 3.9)!
>
> That's unfortunate. So typing.List (and Dict etc.) does not work for >= 3.9 and list[str] does not work for < 3.9.
>
> I am not keen to drop 3.8 support easily (if it can be avoided in some way). If people would want to use this package in the same environment as say deeplabcut, that could cause conflicts.

I'm happy to drop 3.8, or revert to deprecated `typing` hints (which is very uncool)*. Buuuuut if everyone's favourite pose estimation is only stable in 3.8 then that's a compelling reason to swallow the lost cool points.

* [DLC seems to be fine with many python3's](https://github.com/DeepLabCut/DeepLabCut/blob/main/.github/workflows/python-package.yml#L17) - but I'm not really an expert user. What do the researchers use?

*) Technically we could also drop these type hints altogether but that's so uncool it doesn't bear thinking about.

**Comment by niksirbi** [comment]
> * [DLC seems to be fine with many python3's](https://github.com/DeepLabCut/DeepLabCut/blob/main/.github/workflows/python-package.yml#L17) - but I'm not really an expert user. What do the researchers use?

DLC itself maybe cool with many python versions, but what the user ends up with depends on what the package manager resolves to for their environment, right? Particular combinations of `tensorflow` and `CUDA` (which DLC relies on) may force older python versions. But we cannot make everyone happy 🤷🏼‍♂️ 

Out of curiosity I just installed the latest DLC version on my M2 Mac in a fresh conda environment. It ended up with python3.9. 

Maybe we can drop python 3.8 for now, and if the problem does come up in the future, we will know what to do (bring back uncool deprecated type hints)
